src/lib/QueryType.hs

```diff
@@ -181,6 +181,17 @@ naryLamExprType (NaryLamExpr (NonEmptyNest b bs) eff body) = liftTypeQueryM idSu
     binderToPiBinder :: Binder n l -> PiBinder n l
     binderToPiBinder (nameBinder:>ty) = PiBinder nameBinder ty PlainArrow
 
+specializedFunType :: EnvReader m => SpecializationSpec n -> m n (NaryPiType n)
+specializedFunType (AppSpecialization ~(Var f) ab) = liftEnvReaderM $
+  refreshAbs ab \extraArgBs (ListE staticArgs) -> do
+    let extraArgBs' = fmapNest plainPiBinder extraArgBs
+    lookupAtomName (sink f) >>= \case
+      TopFunBound fTy _ -> do
+        NaryPiType dynArgBs effs resultTy <- instantiateNaryPi fTy staticArgs
+        let allBs = fromJust $ nestToNonEmpty $ extraArgBs' >>> nonEmptyToNest dynArgBs
```

Comment:
Is that really guaranteed to be empty? What if somebody decorates a piece of data with `@noinline`?

Reply:
The intention is that `@noinline` is only used for functions. If somebody tries to use it with data they should hit the `Not a valid @noinline function type` error in `execUDecl`. I'll add a test for it though.